### PR TITLE
Significantly faster drop downs for widget.Select and widget.SelectEntry

### LINKED ISSUE #2164: Drop-down menu with many items is slow
### Describe the bug:

Select the drop-down menu year, the corresponding time is too long

### Screenshots:

Peek 2021-04-14 06-44

### Example code:

```go
package main

import (
        "fmt"
        "time"

        "fyne.io/fyne/v2"
        "fyne.io/fyne/v2/app"
        "fyne.io/fyne/v2/container"
        "fyne.io/fyne/v2/layout"
        "fyne.io/fyne/v2/theme"
        "fyne.io/fyne/v2/widget"
)

var t = time.Now().Local()

func main() {
        a := app.New()
        w := a.NewWindow("hello")
        a.Settings().SetTheme(theme.LightTheme())

        //select options
        years := func() []string {
                var s []string
                for i := 1601; i <= 3499; i++ {
                        is := fmt.Sprintf("%dyear", i)
                        s = append(s, is)
                }
                return s
        }

        selecty := widget.NewSelect(years(), func(s string) {
        })
        selecty.PlaceHolder = "year"

        btnOK := widget.NewButton("OK", func() {
                if selecty.SelectedIndex() == -1 {
                        selecty.SetSelectedIndex(t.Year() - 1601)
                }

                fmt.Printf("%s year index=%d\n", selecty.Selected, selecty.SelectedIndex())
        })

        hbox := container.New(layout.NewHBoxLayout(), selecty, btnOK)

        w.SetContent(container.New(layout.NewVBoxLayout(), hbox))
        w.Resize(fyne.Size{600, 600})
        w.ShowAndRun()

}
```

### Device
 - **OS: Linux** 
 - **Version: VERSION_ID="20210401" PRETTY_NAME="openSUSE Tumbleweed"**
 - **Go version: go1.16.2 linux/amd64**
 - **Fyne version: v2.0.2**

### PR DESCRIPTION
**Pull request by Jacalz**
### Description:

This moves the popup menus for both `widget.Select` and `widget.SelectEntry` to use the list widget for their popups. This has significant improvements when it comes to performance. Firstly, it allows us to set up the contents dynamically without having to loop through all the options and create menu items for them. It also makes scrolling much more efficient thanks to the List widgets clever caching. As a result, popup now show pretty much instantly no matter how many items there are.

This also allowed for easily exposing the Options as a variable for `widget.SelectEntry`.
There are a few issues to iron out but it is in a kind of working state already.

Fixes #2164

fast-select

### Checklist:

- [ ] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

#### Where applicable:

- [x] Public APIs match existing style.

### CODE REVIEW

**Comment by Jacalz** [comment]
The tests are not perfect, but it should be ready for some reviews at least :slightly_smiling_face:

**Review by fpabl0 — CHANGES_REQUESTED** [review]
I really love these changes 🥳. I have just added a little suggestion to make it looking better.

File: widget/select_entry.go [hunk 1]
```diff
@@ -73,40 +77,62 @@ func (e *SelectEntry) MinSize() fyne.Size {
...
-	})
-	dropDownButton.Importance = LowImportance
-	dropDownButton.SetIcon(theme.MenuDropDownIcon())
-	return dropDownButton
+	length := func() int {
+		return len(e.Options)
+	}
+
+	create := func() fyne.CanvasObject {
+		return &Label{}
```
[the remark this answers is not in the document]
Comment (Jacalz) [reply widget/select_entry.go:114]: Yeah that's a good idea. I tried to use the actual menu items, but that caused a lot of panics and errors.

**Comment by andydotxyz** [comment]
This certainly is very cool.
It is a shame that using the menu items was not working.
Is it really outside of the box to consider managing menus internally as collection widgets like a list instead of moving this one usage of menu to a list?

**Comment by andydotxyz** [comment]
P.S. in the video it seems like the drop-down falls short of the height of the UI, is that a glitch of the demo code?

**Comment by Jacalz** [comment]
> Is it really outside of the box to consider managing menus internally as collection widgets like a list instead of moving this one usage of menu to a list?

I considered that, but we would be back to square one with needing to loop through and generate MenuItems. Not doing that is probably that largest source for the improvement.

> P.S. in the video it seems like the drop-down falls short of the height of the UI, is that a glitch of the demo code?

No, that's intentional. I put a maximum limit of 10 items as I figured that it was a sensible start. Still need to account for the separator thickness though.

**Comment by andydotxyz** [comment]
> I considered that, but we would be back to square one with needing to loop through and generate MenuItems. Not doing that is probably that largest source for the improvement.

Not sure I follow you here sorry. I was meaning that the menu widget could add caching, so the loop through items and creating widgets was avoided.

**Comment by Jacalz** [comment]
> Not sure I follow you here sorry. I was meaning that the menu widget could add caching, so the loop through items and creating widgets was avoided.

What I meant was that, making the menu widget use list would only solve the issue of scrolling. We would still have to pass in menuItems, which would mean that we are back to square one on looping through the Options slice.

With that being said though, it is certainly worth a try now that I think about it. I might have been wrong in my research for this issue and if that's the case, the largest cause for the slowdown could have been the creation of widgets.

**Comment by Jacalz** [comment]
Hmm, maybe not. I see menu separators potentially becoming an annoying issue to work around. I think I should stick to try and sort out some stuff with the current implementation.

File: widget/select_entry.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jacalz) [reply widget/select_entry.go:124]: I think this was solved thanks to the change to the padding in the List.

**Comment by Jacalz** [comment]
I had to rebase this unfortunately as there were several conflicts. The list is looking a lot better now that the padding is slimmed down:
better-padding
Tests are not updated, which means that they fail. I would still appreciate if some re-reviewing can take place and the n I'll update the tests once there is some feedback on this iteration.

**Review by fpabl0 — COMMENTED** [review]
With the padless collection PR looks better, but I still think we should create a custom item or maybe use just a canvas.Text? Because it still looks bigger than before:

Captura de Pantalla 2021-05-20 a la(s) 12 10 14

Captura de Pantalla 2021-05-20 a la(s) 12 10 31

Also, I don't know if it is possible, but should be great if the popup menu list does not have the horizontal padding.

**Comment by Jacalz** [comment]
Hmm, I'm not quite sure that it should be any smaller. I like the fact that each item is about as big as the Select itself. Smaller menus work great for the entry popup menu, but I think this looks better as it is.

**Comment by fpabl0** [comment]
Fair enough. I have just compared it with other apps, and it seems to be smaller:

Captura de Pantalla 2021-05-20 a la(s) 14 38 08

Captura de Pantalla 2021-05-20 a la(s) 14 37 46

But that's ok, it looks good anyway :), the only thing I would remove is the horizontal padding, as it makes it look unnatural IMHO.

**Comment by Jacalz** [comment]
It was decided on the call yesterday to instead use caching inside the menu directly. Closing for now.

**Pull request closed without merging**